[WIP] Chrome custome tabs

yakirn opened the pull request:
Hi
I think this plugin is a good alternative for the InAppBrowser plugin in terms of UI and UX, and I thought why not give Android users some of that awesomeness with [Chrome Custom Tabs](https://developer.chrome.com/multidevice/android/customtabs) :)

> As of Chrome 45, Chrome Custom Tabs is now generally available to all users of Chrome, on all of Chrome's supported Android versions (Jellybean onwards).

There is still (a lot of) work to do on this PR, but I wanted to open it now in order to start a discussion.
Is this something you see fit in this plugin? If yes, I think we will need to change it's name ;)
The link above lists many advantagas of Custom Tabs over WebView. some of them, like [Warm up](https://developer.chrome.com/multidevice/android/customtabs#warm-up chrome to make pages load faster) or the [Custom Action button](https://developer.chrome.com/multidevice/android/customtabs#configure-a custom action button) requires us to extend this plugin's current set of APIs to allow developers take advantage of. does this seems relevant to you?

Work to be done:
- [x] Really implement the `isAvailable ` function by trying to connect to Chrome's service
- [ ] accept parametrs like transition animation, color and others in the `show` function
- [ ] implement `hide` so that cordova's javacscript code can close the tab (not sure this can be fully supported)
- [x] Implement a `warmUp` function
- [ ] accept action buttons argument in `show` (optional)
- [ ] Did I miss anything?

EddyVerbruggen commented:
Hey @yakirn I love it! Please continue your work. Once `isAvailable` works correctly we can already merge your code so at least devs can already decide if they like it better than inappbrowser. And indeed, a name change would be appropriate once the Android side has matured a bit more. ❤️ 👍

yakirn commented:
Hi @EddyVerbruggen I've implemented `isAvailable`. I also implemented `connectToService` and `warmUp` on the way since I needed to test that the connection actually work.
Only tested it on my device though (Nexus 5 running Marshmallow), according to the docs it should work, but it'll probably be a good idea to test it on a device that don't have chrome on it, just to make sure. Unless you have comments on the code, I think we're ready to merge let other devs test it :)

johnthepink commented:
I've got a couple devices I will test on soon. Thanks for doing this :)

yakirn commented:
Sure @johnthepink, Thanks! :)

EddyVerbruggen commented:
@yakirn merged to master! Let's let others also test this baby. Awesome work!

 If this causes trouble for anyone please use the latest released version instead of master.

ezazwar1 commented:
@yakirn  please update the docs

yakirn commented:
@ezazwar1 oh yes, of course, hopefully later today.
@EddyVerbruggen should I just add it to the commit? if we'll merge this, it might confuse users of the stable version. I can add it to my fork, but then ppl will probably miss it.
for now:
`show` behaves the same in android as in iOS, except that it only accepts url, and will ignore any other option parameter:
`show(successCallback, failCallback, {url: url})`
In android, you can optionally call the following 2 methods sometime before `show` to get better performance / UX:

1. call `connectToService` somewhere in your app.
2. whenever there's a chance the user will open an external url, call `warmUp`

both methods don't accept any arguments besides success and fail callbacks:
`connectToService(successCallback, failCallback)`
`warmUp(successCallback, failCallback)`

I suggest that you will wrap them with a promise for a more convenient use.
You also should call `isAvailable` to check if Chrome custom tabs are supported in the device before calling any other method.
please share your experience using plugin :)

ezazwar1 commented:
Thanks @yakirn  Great Work

EddyVerbruggen commented:
Hey @yakirn I just added you as a committer to this repo so feel free to just clone this one and push anything you like to master. I trust you to not break anything for iOS in the process :)

Thanks again,
Eddy

ezazwar1 commented:
Hey @EddyVerbruggen  Thanks 

i suggest @yakirn  to make his own repo because the name doesn't fit the cause of the plugin and @EddyVerbruggen  you can put a link in this repo to @yakirn  repo

and @EddyVerbruggen  i have few issues in wkwebview if you can have a look at them please

Regards 
Ezaz

EddyVerbruggen commented:
@ezazwar1 suggestions are welcome but you're not in charge of my schedule afaik.

@yakirn and I will think of a strategy as we move forward (name change perhaps). It still makes a lot of sense to me to agree on a uniform API for using an advanced inappbrowser and what better way than having both platform implementations in the same repo?

ezazwar1 commented:
@EddyVerbruggen sorry if you felt bad. Its just a suggestion

davidefavia commented:
Thanks @yakirn, nice job (and of course big thanks to @EddyVerbruggen)!
I tested on my Nexus5 with Android M and it works fine.
Just a note on Cordova build process (for future documentation/troubleshooting section or for anyone experiencing the same problem): I'm using Cordova `6.2.0` with cordova-android `5.1.1` and the build failed with the following error

```
[...]/platforms/android/src/org/chromium/customtabsclient/shared/ServiceConnection.java:32: error: diamond operator is not supported in -source 1.6
        mConnectionCallback = new WeakReference<>(connectionCallback);
```

since my `build.gradle` file had the following default options:

```
compileOptions {
    sourceCompatibility JavaVersion.VERSION_1_6
    targetCompatibility JavaVersion.VERSION_1_6
}
```


I had to add a `build-extras.gradle` (see [Cordova > Android > Project configuration](https://cordova.apache.org/docs/en/latest/guide/platforms/android/#project-configuration)) file in the `platforms/android` folder with the following options to override default behaviour:

```
ext.postBuildExtras = {
    android {
        compileOptions {
            sourceCompatibility JavaVersion.VERSION_1_7
            targetCompatibility JavaVersion.VERSION_1_7
        }
    }
}
```

Thank for the hard work and for future parameters update (can't wait for the `color` one 😉 ).

yakirn commented:
Thanks @davidefavia glad to hear you like it!
Also, thanks for your advice about that java error (I just let Android Studio change the build file for me, and forgot about it. So you didn't miss anything).

In order to really fix the java version issue I made a copy of your `build-extras.gradle` file and reference it from the plugin. This way it should be added automatically to the android project.
I'm also not an android expert, so if anyone has a better idea feel free to share.
I also added an option to change the color of the browser toolbar :)
I hope to be able to push this today so you guys can test it and see if the build fix really work (Maybe you would need to do `cordova platform remove android && cordova platform add android`)

My plan is to create a WIKI page with all the needed information to use this version of the plugin, and link to it from the `README.md` file. this way it won't confuse users of the stable version, and be more accessible. Hows that sounds @EddyVerbruggen ?

vivekpd15 commented:
Great job @yakirn. This works great on my Moto G 1st Gen running Android 5.1. I was wondering if you have any plans to add an exit event listener (when the custom tab is closed) to the code. It would be very nifty to have. Thanks for the great add.

yakirn commented:
Just pushed a new version to master! new features:

1. new function: `mayLaunchUrl(url, success, fail)`
2. support for changing the toolbar color (pass a color param to `show` same as the iOS version of the plugin)
supported color formats can be found [here](https://developer.android.com/reference/android/graphics/Color.html#parseColor(java.lang.String))

also exit event should work just like in iOS. @vivekpd15 please tell me if it's not working.

yakirn commented:
I started a [WIKI](https://github.com/EddyVerbruggen/cordova-plugin-safariviewcontroller/wiki)! it can probably use some more editing ;)

davidefavia commented:
I tried the new `color` option with a `#RRGGBB` value on Android M and it works, thanks @yakirn also for the wiki!
May I suggest to use the same `tintColor` option as in the iOS version just to have unified parameter names? What do you think about it?

yakirn commented:
Glad to hear @davidefavia . 
You're right, it should be `tintColor` to match iOS. funny I missed that...

vivekpd15 commented:
@yakirn @davidefavia Please correct me if I am wrong but in my understanding, tintColor affects only the button colors on iOS and not the toolbar color whereas in Android this affects the complete toolbar color. If this is actually the case, then maybe keeping the names different will be more beneficial since they serve different purposes.

davidefavia commented:
Yes @vivekpd15, you're right, I missed that. Since the parameters serve different purposes maybe it's better to leave different names!

yakirn commented:
Yep. I guess I missed not missing that haha :)
Also it makes sense to have parts of your app look different in iOS and Android since each use different design principles (i.e Material Design and whatever iOS's is called)
but in order to avoid confusion maybe a better name is needed "Toolbar color" is how it is called in chrome's docs, so maybe `toolbarColor`?

vivekpd15 commented:
That sounds great to me. :+1:

vivekpd15 commented:
@yakirn The 'loaded' event handler seems to work fine on my Moto G but 'closed' is not working for some reason.

yakirn commented:
Thanks @VivekP I see the problem now. I will try to fix it soon.


> On 30 May 2016, at 10:09 PM, VivekP <notifications@github.com> wrote:
> 
> @yakirn <https://github.com/yakirn> The 'loaded' event handler seems to work fine on my Moto G but 'closed' is not working for some reason.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/EddyVerbruggen/cordova-plugin-safariviewcontroller/pull/23#issuecomment-222542489>, or mute the thread <https://github.com/notifications/unsubscribe/AGWeMKzg0MY6c6rmCeGdIUg_jQ2z4K68ks5qGzYHgaJpZM4IY0xz>.
>

yakirn commented:
@vivekpd15 please try with the latest version of master. it should send loaded and closed events now.
I also changed `color` to `toolbarColor`.
CC: @davidefavia

davidefavia commented:
Everything still works on my side (testing `toolbarColor` parameter), thanks!

vivekpd15 commented:
@yakirn Thanks a bunch. The plugin works flawlessly now.

akofman commented:
@EddyVerbruggen After renaming this plugin, don't forget to remove the animals in the description. This is not the safari anymore :p

EddyVerbruggen commented:
but but but... I ❤️ those animals!!!

:tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile::tiger: :elephant: :crocodile:
